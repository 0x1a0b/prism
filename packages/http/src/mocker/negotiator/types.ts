import { INodeExample, INodeExternalExample, ISchema } from '@stoplight/types';

export interface IHttpNegotiationResult {
  code: string;
  mediaType: string;
  example?: INodeExample | INodeExternalExample;
  schema?: ISchema;
}
