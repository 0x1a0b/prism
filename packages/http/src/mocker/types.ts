import { IOperationOptions, Protocol } from '@stoplight/prism-core/mocker/types';

export interface IHttpOperationOptions extends IOperationOptions {
  protocol: Protocol.HTTP;
  // Defaults to the lowest 2xx status code
  status: string;
  // Defaults to application/json
  mediaType: string;
  // Defaults to content type
  example?: string;
  // If true mock back response headers
  headers?: boolean;
}
