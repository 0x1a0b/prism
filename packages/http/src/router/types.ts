import { IHttpOperation } from '@stoplight/types/http';

export type Nullable<T> = T | null;

export interface IMatch {
  resource: IHttpOperation;
  serverMatch: MatchType | null;
  pathMatch: MatchType;
}

export enum MatchType {
  CONCRETE = 'concrete',
  TEMPLATED = 'templated',
  NOMATCH = 'no-match',
}
