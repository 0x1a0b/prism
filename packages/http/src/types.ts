import {
  IFilesystemLoaderOpts,
  IPrism,
  IPrismComponents,
  IPrismConfig,
} from '@stoplight/prism-core';
import { IHttpOperation } from '@stoplight/types';

export type TPrismHttpInstance<LoaderInput = IFilesystemLoaderOpts> = IPrism<
  IHttpOperation,
  IHttpRequest,
  IHttpResponse,
  IHttpConfig,
  LoaderInput
>;

export type TPrismHttpComponents<LoaderInput = IFilesystemLoaderOpts> = Partial<
  IPrismComponents<IHttpOperation, IHttpRequest, IHttpResponse, IHttpConfig, LoaderInput>
>;

// TODO: should be complete | and in the @stoplight/types repo
export type IHttpMethod =
  | 'get'
  | 'put'
  | 'post'
  | 'delete'
  | 'options'
  | 'head'
  | 'patch'
  | 'trace'; // ... etc

export interface IHttpOperationConfig {
  mediaType?: string;
  code?: string;
  exampleKey?: string;
  dynamic?: boolean;
}

export interface IHttpConfig extends IPrismConfig {
  mock: boolean | IHttpOperationConfig;

  security?: {
    // TODO
  };

  validate?: {
    request?:
      | boolean
      | {
          hijack?: boolean;
          headers?: boolean;
          query?: boolean;
          body?: boolean;
        };

    response?:
      | boolean
      | {
          headers?: boolean;
          body?: boolean;
        };
  };
}

export interface IHttpRequest {
  method: IHttpMethod;
  url: {
    baseUrl?: string;
    path: string;
    query?: {
      [name: string]: string;
    };
  };
  headers?: {
    [name: string]: string;
  };
  body?: any;
}

export interface IHttpResponse {
  statusCode: number;
  headers?: {
    [name: string]: string;
  };
  body?: any;
}
