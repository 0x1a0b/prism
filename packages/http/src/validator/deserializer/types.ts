import { ISchema } from '@stoplight/types/schema';

export type DeserializeHttpHeader = (value: string, type: string, explode: boolean) => any;

export interface IHttpHeaderParamStyleDeserializer {
  supports: (style: string) => boolean;
  deserialize: DeserializeHttpHeader;
}

export interface IHttpParamDeserializerRegistry<T extends Function> {
  get(style: string): T | undefined;
}

export interface IHttpParamStyleDeserializer<T> {
  supports: (style: string) => boolean;
  deserialize: T;
}

export type DeserializeHttpQuery = (
  key: string,
  query: {
    [name: string]: string | string[];
  },
  schema: ISchema,
  explode: boolean
) => any;

export interface IHttpQueryParamStyleDeserializer
  extends IHttpParamStyleDeserializer<DeserializeHttpQuery> {
  deserialize: DeserializeHttpQuery;
}
