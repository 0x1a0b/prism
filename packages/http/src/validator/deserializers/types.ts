import { HttpParamStyles, ISchema } from '@stoplight/types';

import { IHttpNameValue, IHttpNameValues } from '../../types';

export interface IHttpParamDeserializerRegistry<Parameters, S = HttpParamStyles> {
  get(style: S): IHttpParamStyleDeserializer<Parameters> | undefined;
}

export interface IHttpParamStyleDeserializer<Parameters, S = HttpParamStyles> {
  supports: (style: S) => boolean;
  deserialize: (name: string, parameters: Parameters, schema: ISchema, explode?: boolean) => any;
}

export type IHttpHeaderParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValue>;
export type IHttpQueryParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValues>;
