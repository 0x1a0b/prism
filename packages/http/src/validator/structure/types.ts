import { IValidation } from '@stoplight/prism-core';

export interface IHttpValidator<Obj, Specs> {
  validate(obj: Obj, specs: Specs[], mediaType?: string): IValidation[];
}
