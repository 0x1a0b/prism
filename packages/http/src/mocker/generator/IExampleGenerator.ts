// @todo S is Schema interface (not defined yet)
export interface IExampleGenerator<S = any> {
  generate(schema: S, mediaType: string): Promise<string>;
}
