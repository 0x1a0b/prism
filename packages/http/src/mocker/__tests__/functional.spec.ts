import { ISchema } from '@stoplight/types/schema';
import * as Ajv from 'ajv';

import { httpOperations, httpRequests } from '../../__tests__/fixtures';
import { JSONSchemaExampleGenerator } from '../generator/JSONSchemaExampleGenerator';
import { HttpMocker } from '../index';

// TODO: Turn examples into test cases -> https://stoplightio.atlassian.net/wiki/spaces/PN/pages/5996560/Prism+Feature+List+draft
describe('http mocker', () => {
  const mocker = new HttpMocker(new JSONSchemaExampleGenerator());

  describe('request is valid', () => {
    describe('HttpOperation contains example', () => {
      test('return lowest 2xx code and match response example to media type accepted by request', async () => {
        const response = await mocker.mock({
          resource: httpOperations[0],
          input: httpRequests[0],
        });

        expect(response).toMatchSnapshot();
      });

      test('return lowest 2xx response and the first example matching the media type', async () => {
        const response = await mocker.mock({
          resource: httpOperations[1],
          input: Object.assign({}, httpRequests[0], {
            data: Object.assign({}, httpRequests[0].data, {
              headers: { 'Content-type': 'application/xml' },
            }),
          }),
        });

        expect(response).toMatchSnapshot();
      });

      describe('the media type requested does not match the example', () => {
        test('return lowest 2xx, first example and matching media type', async () => {
          const response = await mocker.mock({
            resource: httpOperations[0],
            input: Object.assign({}, httpRequests[0], {
              data: Object.assign({}, httpRequests[0].data, {
                headers: { 'Content-type': 'application/xml' },
              }),
            }),
          });

          expect(response).toMatchSnapshot();
        });
      });
    });

    describe('HTTPOperation contain no examples', () => {
      test('return dynamic response', async () => {
        if (!httpOperations[1].responses[0].contents[0].schema) {
          throw new Error('Missing test');
        }

        const ajv = new Ajv();
        const validate = ajv.compile(httpOperations[1].responses[0].contents[0].schema as ISchema);

        const response = await mocker.mock({
          resource: httpOperations[1],
          input: httpRequests[0],
          config: {
            mock: {
              dynamic: true,
            },
          },
        });

        expect(response).toHaveProperty('statusCode', 200);
        expect(response).toHaveProperty('headers', { 'Content-type': 'application/json' });
        expect(validate(JSON.parse(response.body))).toBe(true);
      });
    });
  });

  describe('request is invalid', () => {
    test('returns 400 and static error response', async () => {
      const response = await mocker.mock({
        resource: httpOperations[0],
        input: httpRequests[1],
      });

      expect(response).toMatchSnapshot();
    });

    test('returns 400 and dynamic error response', async () => {
      if (!httpOperations[1].responses[1].contents[0].schema) {
        throw new Error('Missing test');
      }

      const response = await mocker.mock({
        resource: httpOperations[1],
        input: httpRequests[1],
      });

      const ajv = new Ajv();
      const validate = ajv.compile(httpOperations[1].responses[1].contents[0].schema as ISchema);

      expect(validate(JSON.parse(response.body))).toBe(true);
    });
  });
});
