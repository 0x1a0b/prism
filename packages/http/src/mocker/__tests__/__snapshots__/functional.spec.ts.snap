// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`http mocker request is invalid returns 400 and static error response 1`] = `
Object {
  "body": Array [
    Object {
      "message": "error",
    },
  ],
  "headers": Object {
    "Content-type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`http mocker request is valid HttpOperation contains example return lowest 2xx code and match response example to media type accepted by request 1`] = `
Object {
  "body": Array [
    Object {
      "completed": true,
      "id": 1,
      "name": "make prism",
    },
  ],
  "headers": Object {
    "Content-type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`http mocker request is valid HttpOperation contains example return lowest 2xx response and the first example matching the media type 1`] = `
Object {
  "body": "<todo><name>Shopping</name><completed>false</completed></todo>",
  "headers": Object {
    "Content-type": "application/xml",
  },
  "statusCode": 200,
}
`;

exports[`http mocker request is valid HttpOperation contains example the media type requested does not match the example return lowest 2xx, first example and matching media type 1`] = `
Object {
  "body": "{ \\"root\\": \\"first\\" }",
  "headers": Object {
    "Content-type": "application/xml",
  },
  "statusCode": 200,
}
`;

exports[`http mocker request is valid given enforced example key and mediaType should return 200 response 1`] = `
Object {
  "body": "{ \\"root\\": \\"second\\" }",
  "headers": Object {
    "Content-type": "application/xml",
  },
  "statusCode": 200,
}
`;

exports[`http mocker request is valid given enforced example key should return application/json, 200 response 1`] = `
Object {
  "body": Array [
    Object {
      "completed": false,
      "id": 2,
      "name": "make bears",
    },
  ],
  "headers": Object {
    "Content-type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`http mocker request is valid given enforced status code and contentType and exampleKey should return the matching example 1`] = `
Object {
  "body": "<root>second</root>",
  "headers": Object {
    "Content-type": "application/xml",
  },
  "statusCode": 201,
}
`;

exports[`http mocker request is valid given enforced status code and contentType should return the first matching example 1`] = `
Object {
  "body": "<root>first</root>",
  "headers": Object {
    "Content-type": "application/xml",
  },
  "statusCode": 201,
}
`;

exports[`http mocker request is valid given enforced status code and example key should return application/json example 1`] = `
Object {
  "body": "{ \\"root\\": \\"second\\" }",
  "headers": Object {
    "Content-type": "application/json",
  },
  "statusCode": 201,
}
`;

exports[`http mocker request is valid given enforced status code should return the first matching example of application/json 1`] = `
Object {
  "body": "{ \\"root\\": \\"first\\" }",
  "headers": Object {
    "Content-type": "application/json",
  },
  "statusCode": 201,
}
`;
